src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs

```diff
@@ -134,8 +134,14 @@ private static GraphQLFragmentDefinition GetFirstFragmentWithoutPendingDependenc
                 HashSet<GraphQLFragmentDefinition>? dependencies = null;
                 _selectionSetsToVisit.Push(def.SelectionSet);
 
+                int counter = 0;
+                const int MAX_ITERATIONS = 2000;
                 while (_selectionSetsToVisit.Count > 0)
                 {
+                    // https://github.com/graphql-dotnet/graphql-dotnet/issues/3527
+                    if (++counter > MAX_ITERATIONS)
+                        throw new ValidationError("It looks like document has fragment cycle. Please make sure you are using standard validation rules especially NoFragmentCycles one.");
```

Comment:
It would seem that with this check in place there should be no need to add the `PostponeComplexityValidationVisitor` and related code.

Reply:
Besides the fact that the error would be reported twice, that is.

Reply:
I strongly feel that complexity analysis should be run after all standard validation rules if any. Remember that we switched that code into validation rule as a internal techical aspect. One can wrap any code into validation rule, but this does not mean that it can and should be launched before the basic checks are performed.